level/level.c

```diff
@@ -329,73 +333,116 @@ int rec_partition(int **room_map, int x, int y, int w, int h, int rm) {
+                        if (abs(dx+dy) == 1) {
+                            potential_doors[x][y] = true;
+                        }
                     }
                 }
             }
         }
     }
 
-    bool room_accessible[rooms];
-    for (int i = 0; i < rooms; i++) room_accessible[i] = false;
-    room_accessible[partitioning[0][0]] = true;
-    for (int x = 0; x < lvl->width; x++) {
-        for (int y = 0; y < lvl->height; y++) {
+    // whether each room is accessible from the "root" room
+    bool room_connected[max_room_id];
+
+    for (int i = 0; i < max_room_id; i++) {
+        room_connected[i] = false;
+    }
+
+    // the "root" room
+    int rand_x = (rand() % (lvl->width - 1)) + 1;
+    int rand_y = (rand() % (lvl->height - 1)) + 1;
+
+    room_connected[room_ids[rand_x][rand_y]] = true;
+
+    // We are building a tree of connected rooms (via door placement)
+    // starting at a "root" room. Potential doors only become doors if
+    // they would connected an already-connected room to an unconnected
+    // room.
+    //TODO Any problem with skipping the border squares?
+    for (int x = 1; x < lvl->width - 1; x++) {
+        for (int y = 1; y < lvl->height - 1; y++) {
             if (potential_doors[x][y]) {
-                bool door_needed = false;
                 bool door_possible = false;
                 int rm_a, rm_b;
+
                 if (x+1 < lvl->width && x-1 >= 0 && lvl->tiles[x+1][y] != TILE_WALL && lvl->tiles[x-1][y] != TILE_WALL) {
-                    rm_a = partitioning[x+1][y];
-                    rm_b = partitioning[x-1][y];
+                // a "horizontal" door
+                    rm_a = room_ids[x+1][y];
+                    rm_b = room_ids[x-1][y];
                     door_possible = true;
-                }
-                if (y+1 < lvl->height && y-1 >= 0 && lvl->tiles[x][y+1] != TILE_WALL && lvl->tiles[x][y-1] != TILE_WALL) {
-                    rm_a = partitioning[x][y+1];
-                    rm_b = partitioning[x][y-1];
+                } else if (y+1 < lvl->height && y-1 >= 0 && lvl->tiles[x][y+1] != TILE_WALL && lvl->tiles[x][y-1] != TILE_WALL) {
+                // a "vertical" door
+                    rm_a = room_ids[x][y+1];
+                    rm_b = room_ids[x][y-1];
                     door_possible = true;
                 }
-                if (door_possible && rand()%100 > 0) {
-                    if (room_accessible[rm_a] + !room_accessible[rm_b] != 1) {
-                        door_needed = true;
-                        room_accessible[rm_b]=true;
-                        room_accessible[rm_a]=true;
-                    }
+
+                if (! door_possible) {
+                    //DEBUG Checking whether the vertical/horizontal test is redundant
+                    logger("No door possible at (%d,%d)\n", x, y);
+                    //TODO It is not redundant currently. It rejects T-junction wall squares.
```

Comment:
Do we need this check just to catch T-junctions? Is there a clever way to do this up above? Do we care?

Reply:
You could probably be more clever, yes.